frb/halos.py

```diff
@@ -0,0 +1,1016 @@
+
+    # Table the halos
+    halo_tbl = Table()
+    halo_tbl['trial'] = halo_i
+    halo_tbl['M'] = M_i
+    halo_tbl['R'] = R_i
+    halo_tbl['DM'] = DM_i
+    halo_tbl['z'] = z_i
+
+    # Write
+    if outfile is not None:
+        print("Writing to {}".format(outfile))
+        np.save(outfile, DM_grid, allow_pickle=False)
+        halo_tbl.write(outfile+'.fits', overwrite=True)
+
+    return DM_grid, halo_tbl
+
+
+def rad3d2(xyz):
+    """ Calculate radius to x,y,z inputted
+    Assumes the origin is 0,0,0
+
+    Parameters
+    ----------
+        xyz : Tuple or ndarray
+
+    Returns
+    -------
+        rad3d : float or ndarray
+
+    """
+    return xyz[0]**2 + xyz[1]**2 + xyz[-1]**2
+
+
+class ModifiedNFW(object):
+    """ Generate a modified NFW model, e.g. Mathews & Prochaska 2017
+    for the hot, virialized gas.
+
+    Parameters:
+        log_Mhalo: float, optional
+          log10 of the Halo mass (solar masses)
+        c: float, optional
+          concentration of the halo
+        f_hot: float, optional
+          Fraction of the baryons in this hot phase
+          Will likely use this for all diffuse gas
+        alpha: float, optional
+          Parameter to modify NFW profile power-law
+        y0: float, optional
+          Parameter to modify NFW profile position
+        z: float, optional
+          Redshift of the halo
+
+    Attributes:
+        H0: Quantity;  Hubble constant
+        fb: float; Cosmic fraction of baryons (stars+dust+gas) in the entire halo
+           Default to 0.16
+        r200: Quantity
+           Virial radius
+        rho0: Quantity
+           Density normalization
+        M_b: Quantity
+           Mass in baryons of the
+
+
+    """
+    def __init__(self, log_Mhalo=12.2, c=7.67, f_hot=0.75, alpha=0., y0=1., z=0., **kwargs):
+        # Init
+        # Param
+        self.log_Mhalo = log_Mhalo
+        self.M_halo = 10.**self.log_Mhalo * constants.M_sun.cgs
+        self.c = c
+        self.alpha = alpha
+        self.y0 = y0
+        self.z = z
+        self.f_hot = f_hot
+        self.zero_inner_ne = 0. # kpc
+
+        # Init more
+        self.setup_param()
+
+    def setup_param(self, cosmo=None):
+        """ Setup key parameters of the model
+        """
+        # Cosmology
+        if cosmo is None:
+            self.rhoc = 9.2e-30 * units.g / units.cm**3
+            self.fb = 0.16       # Baryon fraction
+            self.H0 = 70. *units.km/units.s/ units.Mpc
+        else:
+            self.rhoc = cosmo.critical_density(self.z)
+            self.fb = cosmo.Ob0/cosmo.Om0
+            self.H0 = cosmo.H0
+        # Dark Matter
+        self.r200 = (((3*self.M_halo) / (4*np.pi*200*self.rhoc))**(1/3)).to('kpc')
+        self.rho0 = 200*self.rhoc/3 * self.c**3 / self.fy_dm(self.c)   # Central density
+        # Baryons
+        self.M_b = self.M_halo * self.fb
+        self.rho0_b = (self.M_b / (4*np.pi) * (self.c/self.r200)**3 / self.fy_b(self.c)).cgs
```

Comment:
Is there a 3 missing in this mass to density calculation?

Reply:
Good to worry about this.
But I've double checked it is correct.